src/ring_theory/graded_algebra/radical.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import ring_theory.graded_algebra.homogeneous_ideal
+
+/-!
+
+This file contains a proof that the radical of any homogeneous ideal is a homogeneous ideal
+
+## Main statements
+
+* `homogeneous_ideal.is_prime_iff`: a homogeneous ideal `I` is prime if and only if `I` is
+  homogeneously prime, i.e. if `x, y` are homogeneous elements such that `x * y ∈ I`, then
+  at least one of `x,y` is in `I`.
+* `homogeneous_ideal.rad`: radical of homogeneous ideal is a homogeneous ideal.
+
+## Implementation details
+
+Through out this file, the indexing type `ι` of grading is assumed to be a
+`linear_ordered_cancel_add_comm_monoid`. This might be stronger than necessary and `linearith`
+does not work on `linear_ordered_cancel_add_comm_monoid`.
+
+## Tags
+
+homogeneous, radical
+-/
+
+
+open_locale big_operators
+
+section linear_ordered_cancel_add_comm_monoid
+
+variables {ι R A : Type*}
+variables [comm_semiring R] [comm_ring A] [algebra R A]
+variables [linear_ordered_cancel_add_comm_monoid ι] [decidable_eq ι]
+variables {𝒜 : ι → submodule R A} [graded_algebra 𝒜]
+
+lemma ideal.is_homogeneous.is_prime_of_homogeneous_mem_or_mem
+  {I : ideal A} (hI : I.is_homogeneous 𝒜) (I_ne_top : I ≠ ⊤)
+  (homogeneous_mem_or_mem : ∀ {x y : A},
+    set_like.is_homogeneous 𝒜 x → set_like.is_homogeneous 𝒜 y
+    → (x * y ∈ I → x ∈ I ∨ y ∈ I)) : ideal.is_prime I :=
+⟨I_ne_top, begin
```

Comment:
I still find it quite hard to follow what the "core" of the proof is in this `begin` / `end` block - could you add a comment within the proof explaining the overall strategy? In particular, it's not at all clear to me whether `linear_ordered_cancel_add_comm_monoid` is necessary.

Reply:
I have added some documentation.

Reply:
> In particular, it's not at all clear to me whether linear_ordered_cancel_add_comm_monoid is necessary for this to be true, or just necessary for this particular proof strategy.

I have actually discussed this point with @kbuzzard once, If I remembered correctly, I think we agreed that this might be necessary in general.

Reply:
This might be a counter example:

consider `ℤ ⨁ ℤ` and give `(x,0)` grade 0 and `(0, y)` grade 1. Then ideal generated by `(6,6)` is not prime, but it satisfies `homogeneous_mem_or_mem` condition.

Reply:
I don't think that's a valid grading, what grade is `(0, 4) = (0,2)*(0,2)`?

Reply:
We could have given the indexing set a weird addition: 0 + 0 = 0, 0 + 1 = 1, 1 + 1 = 1 (I hope this isn't contradictory)

Reply:
Here is an explicit counterexample @eric-wieser 

```lean
import ring_theory.graded_algebra.homogeneous_ideal

inductive two
| z
| o

open two
namespace test

lemma two.o_ne_z : o ≠ z := by tauto
lemma two.z_ne_o : z ≠ o := by tauto

instance has_add : has_add two :=
{ add := λ m n,
  match m, n with
  | z, z := z
  | z, o := o
  | o, z := o
  | o, o := o
  end }

instance : add_comm_monoid two :=
{ zero := z,
  add_comm := λ a b, begin
    cases a; cases b;
    trivial,
  end,
  add_zero := λ a, begin
    cases a; trivial,
  end,
  zero_add := λ a, begin
    cases a; trivial,
  end,
  add_assoc := λ a b c, begin
    cases a; cases b; cases c;
    refl,
  end,
  ..test.has_add}

@[derive [comm_ring]]
def Z_sq : Type := ℤ × ℤ

def submodule_z : submodule ℤ Z_sq :=
{ carrier := { zz | zz.1 = zz.2 },
  zero_mem' := by simp,
  add_mem' := λ a b ha hb, begin
    change a.1 = a.2 at ha,
    change b.1 = b.2 at hb,
    change a.1 + b.1 = a.2 + b.2,
    rw [ha, hb],
  end,
  smul_mem' := λ a b hb, begin
    change a * b.1 = a * b.2,
    change b.1 = b.2 at hb,
    rw [hb],
  end }

def submodule_o : submodule ℤ Z_sq :=
{ carrier := { zz | zz.1 = 0 },
  zero_mem' := by simp,
  add_mem' := λ a b ha hb, begin
    change a.1 = 0 at ha,
    change b.1 = 0 at hb,
    change a.1 + b.1 = 0,
    rw [ha, hb, zero_add],
  end,
  smul_mem' := λ a b hb, begin
    change b.1 = 0 at hb,
    change a * b.1 = 0,
    rw [hb, mul_zero],
  end }

def grading : two → submodule ℤ Z_sq
| z := submodule_z
| o := submodule_o

variable [decidable_eq two]

lemma grading.one_mem : (1 : Z_sq) ∈ grading 0 :=
begin
  change ((1, 1) : Z_sq).1 = ((1, 1) : Z_sq).2,
  refl,
end

lemma grading.mul_mem ⦃i j : two⦄ {a b : Z_sq} (ha : a ∈ grading i) (hb : b ∈ grading j) :
  a * b ∈ grading (i + j) :=
begin
  cases i; cases j,
  { change _ ∈ grading z,
    change a.1 = a.2 at ha,
    change b.1 = b.2 at hb,
    change a.1 * b.1 = a.2 * b.2,
    rw [ha, hb], },
  { change _ ∈ grading o,
    change a.1 = a.2 at ha,
    change b.1 = 0 at hb,
    change a.1 * b.1 = 0,
    rw [hb, mul_zero], },
  { change _ ∈ grading o,
    change a.1 = 0 at ha,
    change a.1 * b.1 = 0,
    rw [ha, zero_mul], },
  { change _ ∈ grading o,
    change a.1 = 0 at ha,
    change a.1 * b.1 = 0,
    rw [ha, zero_mul], },
end

def grading.decompose : Z_sq →+ direct_sum two (λ i, grading i) :=
{ to_fun := λ zz, direct_sum.of (λ i, grading i) z ⟨(zz.1, zz.1), rfl⟩ + 
    direct_sum.of (λ i, grading i) o ⟨(0, zz.2 - zz.1), rfl⟩,
  map_zero' := begin
    simp only [show (0 : Z_sq).1 = 0, from rfl, show (0 : Z_sq).2 = 0, from rfl],
    ext; 
    cases i,
    { simp only [direct_sum.zero_apply, direct_sum.add_apply, map_add, direct_sum.of_eq_same],
      rw direct_sum.of_eq_of_ne,
      refl,
      exact two.o_ne_z, },
    { simp only [direct_sum.zero_apply, direct_sum.add_apply, map_add, direct_sum.of_eq_same],
      rw direct_sum.of_eq_of_ne,
      refl,
      exact two.z_ne_o, },
    { simp only [direct_sum.zero_apply, direct_sum.add_apply, map_add, direct_sum.of_eq_same],
      rw direct_sum.of_eq_of_ne,
      refl,
      exact two.o_ne_z, },
    { simp only [direct_sum.zero_apply, direct_sum.add_apply, map_add, direct_sum.of_eq_same],
      rw direct_sum.of_eq_of_ne,
      refl,
      exact two.z_ne_o, },
  end,
  map_add' := λ zz1 zz2, begin
    cases zz1 with a1 b1,
    cases zz2 with a2 b2,
    simp only [show ∀ (c d : Z_sq), (c + d).1 = c.1 + d.1, from λ _ _, rfl, 
      show ∀ (c d : Z_sq), (c + d).2 = c.2 + d.2, from λ _ _, rfl],
    ext; 
    cases i,
    { simp only [direct_sum.add_apply, map_add, direct_sum.of_eq_same],
      rw [direct_sum.of_eq_of_ne, add_zero, direct_sum.of_eq_of_ne, direct_sum.of_eq_of_ne, 
        add_zero, add_zero],
      change a1 + a2 = a1 + a2,
      refl,
      all_goals { exact two.o_ne_z }, },
    { simp only [direct_sum.add_apply, map_add, direct_sum.of_eq_same],
      rw [direct_sum.of_eq_of_ne, zero_add, direct_sum.of_eq_of_ne, direct_sum.of_eq_of_ne, 
        zero_add, zero_add],
      change (0 : ℤ) = 0 + 0,
      rw zero_add,
      all_goals { exact two.z_ne_o }, },
    { simp only [direct_sum.add_apply, map_add, direct_sum.of_eq_same],
      rw [direct_sum.of_eq_of_ne, add_zero, direct_sum.of_eq_of_ne, direct_sum.of_eq_of_ne, 
        add_zero, add_zero],
      change a1 + a2 = a1 + a2,
      refl,
      all_goals { exact two.o_ne_z }, },
    { simp only [direct_sum.add_apply, map_add, direct_sum.of_eq_same],
      rw [direct_sum.of_eq_of_ne, zero_add, direct_sum.of_eq_of_ne, direct_sum.of_eq_of_ne, 
        zero_add, zero_add],
      change b1 + b2 - (a1 + a2) = (b1 - a1) + (b2 - a2),
      ring,
      all_goals { exact two.z_ne_o }, },
  end }

lemma grading.decompose_z (zz : Z_sq) :
  (grading.decompose zz) z = ⟨(zz.1, zz.1), rfl⟩ :=
begin
  cases zz with a b,
  unfold grading.decompose,
  simp only [add_monoid_hom.coe_mk, direct_sum.add_apply, direct_sum.of_eq_same, add_right_eq_self],
  rw direct_sum.of_eq_of_ne,
  exact two.o_ne_z,
end

lemma grading.decompose_o (zz : Z_sq) :
  (grading.decompose zz) o = ⟨(0, zz.2 - zz.1), rfl⟩ :=
begin
  cases zz with a b,
  unfold grading.decompose,
  simp only [add_monoid_hom.coe_mk, direct_sum.add_apply, direct_sum.of_eq_same, add_left_eq_self],
  rw direct_sum.of_eq_of_ne,
  exact two.z_ne_o,
end

lemma grading.left_inv : 
  function.left_inverse grading.decompose (direct_sum.submodule_coe grading) := λ zz,
begin
  induction zz using direct_sum.induction_on with i zz d1 d2 ih1 ih2,
  { simp only [map_zero],},
  { cases i,
    { rcases zz with ⟨⟨a, b⟩, (rfl : a = b)⟩,
      ext; cases i,
      { rw grading.decompose_z,
        simp, },
      { rw grading.decompose_o,
        simp only [subtype.coe_mk], 
        rw direct_sum.of_eq_of_ne,
        refl,
        exact two.z_ne_o, },
      { rw grading.decompose_z,
        simp, },
      { rw grading.decompose_o,
        simp only [direct_sum.submodule_coe_of, subtype.coe_mk, sub_self], 
        rw direct_sum.of_eq_of_ne,
        refl,
        exact two.z_ne_o, }, },
    { rcases zz with ⟨⟨a, b⟩, (rfl : a = 0)⟩,
      ext; cases i,
      { rw grading.decompose_z,
        simp only [direct_sum.submodule_coe_of, subtype.coe_mk, sub_self], 
        rw direct_sum.of_eq_of_ne,
        refl,
        exact two.o_ne_z, },
      { rw grading.decompose_o,
        simp, },
      { rw grading.decompose_z,
        simp only [direct_sum.submodule_coe_of, subtype.coe_mk, sub_self], 
        rw direct_sum.of_eq_of_ne,
        refl,
        exact two.o_ne_z, },
      { rw grading.decompose_o,
        simp, }, } },
  { simp only [map_add, ih1, ih2], },
end

lemma grading.right_inv : 
  function.right_inverse grading.decompose (direct_sum.submodule_coe grading) := λ zz,
begin
  cases zz with a b,
  unfold grading.decompose,
  simp only [add_monoid_hom.coe_mk, map_add, direct_sum.submodule_coe_of, subtype.coe_mk, prod.mk_add_mk, add_zero,
  add_sub_cancel'_right],
end

instance : graded_algebra grading :=
{ one_mem := grading.one_mem,
  mul_mem := grading.mul_mem,
  decompose' := grading.decompose,
  left_inv := grading.left_inv,
  right_inv := grading.right_inv }

def I : ideal Z_sq := ideal.span {((6, 0) : Z_sq)}

lemma homogeneous_mem_or_mem {x y : Z_sq} (hx : set_like.is_homogeneous grading x)
  (hy : set_like.is_homogeneous grading y) 
  (hxy : x * y ∈ I) : x ∈ I ∨ y ∈ I :=
begin
unfold I at hxy ⊢,
rw ideal.mem_span_singleton at hxy,
rcases hxy with ⟨⟨a, b⟩, hz⟩,
simp only [prod.mk_mul_mk] at hz,
rw prod.eq_iff_fst_eq_snd_eq at hz,
have eq1 : x.1 * y.1 = 6 * a := hz.1,
have eq2 : x.2 * y.2 = 0 * b := hz.2,  
rcases hx with ⟨(_|_), ha⟩;
rcases hy with ⟨(_|_), hb⟩,
{ change x.1 = x.2 at ha,
  change y.1 = y.2 at hb,
  rw [zero_mul, ← ha, ← hb, mul_eq_zero] at eq2,
  cases eq2,
  { left,
    have eq3 : x = 0,
    { ext,
      exact eq2,
      rw ← ha,
      exact eq2, },
    rw eq3,
    exact submodule.zero_mem _, },
  { right,
    have eq3 : y = 0,
    { ext,
      exact eq2, 
      rw ← hb,
      exact eq2, },
    rw eq3,
    exact submodule.zero_mem _, }, },
{ change x.1 = x.2 at ha,
  change y.1 = 0 at hb,
  rw [zero_mul, mul_eq_zero] at eq2,
  cases eq2,
  { left,
    rw ideal.mem_span_singleton,
    use (0, 0),
    ext,
    { erw [ha, eq2, mul_zero],
      refl, },
    { erw [eq2, mul_zero],
      refl, }, },
  { right,
    rw ideal.mem_span_singleton,
    use (0, 0),
    ext,
    { erw [hb, mul_zero],
      refl, },
    { erw [eq2, mul_zero],
      refl, }, } },
{ change y.1 = y.2 at hb,
  change x.1 = 0 at ha,
  rw [zero_mul, mul_eq_zero] at eq2,
  cases eq2,
  { left,
    rw ideal.mem_span_singleton,
    use (0, 0),
    ext,
    { erw [ha, mul_zero],
      refl, },
    { erw [eq2, mul_zero],
      refl, }, },
  { right,
    rw ideal.mem_span_singleton,
    use (0, 0),
    ext,
    { erw [hb, mul_zero, eq2],
      refl, },
    { erw [eq2, mul_zero],
      refl, }, }, },
{ change x.1 = 0 at ha,
  change y.1 = 0 at hb,
  rw [zero_mul, mul_eq_zero] at eq2,
  cases eq2,
  { left,
    rw ideal.mem_span_singleton,
    use (0, 0),
    ext,
    { erw [ha, mul_zero],
      refl, },
    { erw [eq2, mul_zero],
      refl, }, },
  { right,
    rw ideal.mem_span_singleton,
    use (0, 0),
    ext,
    { erw [hb, mul_zero],
      refl, },
    { erw [eq2, mul_zero],
      refl, }, }, }
end

end test
```lean

Reply:
Here `{(a, a)}` has grade 0 and `{(0, b)}` has grade 1. addition is : `0 + 0 = 0, _ + _ = 1`. The counter example is ideal spaned by `(6,0)`. This is not prime, but satisfies `homogeneous_mem_or_mem`

Reply:
That looks worth contributing to the `counterexamples` directory once this PR is merged.